File: bestia-models/src/main/java/net/bestia/model/Bestia.java
package net.bestia.model;

import javax.persistence.AttributeOverride;
import javax.persistence.AttributeOverrides;
import javax.persistence.Column;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

@Entity
@Inheritance(strategy = InheritanceType.JOINED)
public class Bestia {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	@Column(name = "bestia_id")
	private int id;

	@Embedded
	@AttributeOverrides({
			@AttributeOverride(name = "hp", column = @Column(name = "bHp")),
			@AttributeOverride(name = "mana", column = @Column(name = "bMana")),
			@AttributeOverride(name = "atk", column = @Column(name = "bAtk")),
			@AttributeOverride(name = "def", column = @Column(name = "bDef")),
			@AttributeOverride(name = "spAtk", column = @Column(name = "bSpAtk")),
			@AttributeOverride(name = "spDef", column = @Column(name = "bSpDef")),
			@AttributeOverride(name = "spd", column = @Column(name = "bSpd")) })
	@JsonIgnore
	private BaseValues baseValue;
	@JsonProperty("bdbn")
	private String databaseName;
	@Enumerated(EnumType.STRING)
	@JsonProperty("ele")
	private Element element;
	@JsonProperty("img")
	private String image;
	@JsonProperty("s")
	private String sprite;



	public String getDatabaseName() {
		return databaseName;
	}


	public String getImage() {
		return image;
	}


	public String getSprite() {
		return sprite;
	}
	
	public Element getElement() {
		return element;
	}



	/**
	 * Sets the attacks of a bestia. A bestia can not learn more then 4 attacks
	 * + one attack which does only heal/buff effects.
	 * 
	 * @param attacks
	 */
	/*
	public void setAttacks(Set<Attack> attacks) {
		boolean hasBuff = false;
		for (Attack atk : attacks) {
			if (atk.getBasedStatus() == AttackBasedStatus.NO_DAMAGE) {
				hasBuff = true;
				break;
			}
		}
		if ((hasBuff && attacks.size() > 5) || (!hasBuff && attacks.size() > 4)) {
			throw new IllegalArgumentException(
					"Bestia can only learn 4 attacks. Or 4 Attacks + 1 NO_DAMAGE Attack.");
		}
		this.attacks = attacks;
	}*/
}

File: bestia-models/src/main/java/net/bestia/model/domain/Bestia.kt
package net.bestia.model.domain

import java.io.Serializable

import javax.persistence.AttributeOverride
import javax.persistence.AttributeOverrides
import javax.persistence.Column
import javax.persistence.Embedded
import javax.persistence.Entity
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.Id
import javax.persistence.PrimaryKeyJoinColumn
import javax.persistence.Table
import javax.persistence.Transient

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty

@Entity
@Table(name = "bestias")
class Bestia(
    @Id
    @JsonIgnore
    val id: Int = 0,

    /**
     * The database name.
     *
     * @return The database name.
     */
    @Column(name = "bestia_db_name", unique = true, nullable = false, length = 100)
    @JsonProperty("bdbn")
    val databaseName: String,

    @Column(name = "default_name", nullable = false, length = 100)
    @JsonIgnore
    var defaultName: String,

    @Enumerated(EnumType.STRING)
    @JsonProperty("ele")
    val element: Element,

    @JsonProperty("img")
    val image: String,

    @JsonProperty("sp")
    @AttributeOverrides(AttributeOverride(name = "type", column = Column(name = "visualType")))
    val spriteInfo: SpriteInfo,

    /**
     * Experience points gained if bestia was defeated.
     *
     * @return
     */
    @JsonIgnore
    val expGained: Int,

    /**
     * Returns the type of the bestia.
     *
     * @return The type of the bestia.
     */
    @Enumerated(EnumType.STRING)
    @JsonProperty("t")
    val type: BestiaType,

    @JsonIgnore
    val level: Int,

    @JsonIgnore
    val isBoss: Boolean = false,

    /**
     * Returns the status points of this bestia.
     *
     * @return The status points.
     */
    @JsonIgnore
    @Transient
    val statusPoints: StatusPoints,

    @Embedded
    @JsonIgnore
    val baseValues: BaseValues,

    /**
     * Script which will be attached to this bestia.
     */
    @JsonIgnore
    val scriptExec: String? = null
) : Serializable {

  override fun toString(): String {
    return String.format("Bestia[db: %s, id: %d, level: %d]", databaseName, id, level)
  }
}
